### PR TITLE
[api] fix getTransactionReceipt

### PR DESCRIPTION
**Pull request by Liuhaai**
`null` should be returned when no receipt or transaction is found

### CODE REVIEW

File: api/web3server.go [hunk 1]
```diff
@@ -91,12 +95,13 @@ var (
...
-	errNotImplemented = errors.New("method not implemented")
-	errInvalidFiterID = errors.New("filter not found")
-	errInvalidBlock   = errors.New("invalid block")
+	errUnkownType           = errors.New("wrong type of params")
+	errNullPointer          = errors.New("null pointer")
+	errInvalidFormat        = errors.New("invalid format of request")
+	errNotImplemented       = errors.New("method not implemented")
+	errInvalidFiterID       = errors.New("filter not found")
+	errInvalidBlock         = errors.New("invalid block")
+	errNotFound       error = nil // null is returned when no result was found
```
Comment (dustinxie) [inline api/web3server.go:104]: `errEmptyResult` is more proper name?

File: api/web3server.go [hunk 2]
```diff
@@ -160,13 +165,13 @@ func (svr *Web3Server) handlePOSTReq(req *http.Request) interface{} {
...
-	web3Resps := make([]web3Resp, 0)
+	web3Resps := make([]interface{}, 0)
 	for _, web3Req := range web3Reqs {
 		var (
-			res    interface{}
-			err    error
-			params = web3Req.Get("params").Value()
-			method = web3Req.Get("method").Value()
+			res    interface{} = nil
+			err    error       = nil
```
Comment (dustinxie) [inline api/web3server.go:172]: = errEmptyResult

File: api/web3server.go [hunk 3]
```diff
@@ -742,21 +763,26 @@ func (svr *Web3Server) getTransactionByBlockNumberAndIndex(in interface{}) (inte
...
-	log.L().Error("acquire block meta")
 	blkMetas, err := svr.coreService.BlockMetas(num, 1)
-	if err != nil {
-		return nil, err
-	}
-	if len(blkMetas) == 0 {
-		return nil, errInvalidBlock
+	if err != nil || len(blkMetas) == 0 {
+		if err != nil {
+			log.L().Debug("err in getTransactionByBlockNumberAndIndex", zap.Error(err))
```
Comment (dustinxie) [inline api/web3server.go:769]: why combine these 2 again?

File: api/web3server.go [hunk 4]
```diff
@@ -245,12 +250,13 @@ func (svr *Web3Server) handlePOSTReq(req *http.Request) interface{} {
...
-		if err != nil || !isResultValid(resp) {
+		if err != nil {
 			// temporally used for monitor and debug
-			log.L().Error("web3 server err", zap.String("input", fmt.Sprintf("%+v", web3Req)), zap.Error(err))
+			log.L().Error("web3 server err",
+				zap.String("input", fmt.Sprintf("%+v", web3Req)),
+				zap.Error(err))
 		}
-		web3Resps = append(web3Resps, resp)
+		web3Resps = append(web3Resps, packAPIResult(res, err, int(web3Req.Get("id").Int())))
```
Comment (dustinxie) [inline api/web3server.go:259]: packAPIResult may return a `web3Err` object, in that case would append work?

File: api/web3server.go [hunk 5]
```diff
@@ -337,13 +343,12 @@ func (svr *Web3Server) getBlockByNumber(in interface{}) (interface{}, error) {
...
 
 	blkMetas, err := svr.coreService.BlockMetas(num, 1)
-	if err != nil {
-		return nil, err
-	}
-	if len(blkMetas) == 0 {
-		return nil, errInvalidBlock
+	if err != nil || len(blkMetas) == 0 {
+		if err != nil {
+			log.L().Debug("err in getBlockByNumber", zap.Error(err))
```
Comment (dustinxie) [inline api/web3server.go:348]: why combine again?
`svr.coreService.BlockMetas()` code is not changed, i think only need to add the `log.L().Debug()` line?

File: api/web3server.go [hunk 6]
```diff
@@ -337,13 +343,12 @@ func (svr *Web3Server) getBlockByNumber(in interface{}) (interface{}, error) {
 	}
 
 	blkMetas, err := svr.coreService.BlockMetas(num, 1)
-	if err != nil {
-		return nil, err
-	}
-	if len(blkMetas) == 0 {
-		return nil, errInvalidBlock
```
Comment (dustinxie) [inline api/web3server.go:344]: after this PR, will `errInvalidBlock` still be used? if not, remove it

File: api/web3server.go (hunk 3 above)
Comment (dustinxie) [inline api/web3server.go:769]: same here

File: api/coreservice.go [hunk 7]
```diff
@@ -1036,11 +1036,14 @@ func (core *coreService) BlockMetas(start uint64, count uint64) ([]*iotextypes.B
...
 
-	tipHeight := core.bc.TipHeight()
+	var (
+		tipHeight = core.bc.TipHeight()
+		res       []*iotextypes.BlockMeta
+	)
 	if start > tipHeight {
-		return nil, status.Error(codes.InvalidArgument, "start height should not exceed tip height")
+		log.L().Debug("err in BlockMetas()", zap.Error(status.Error(codes.InvalidArgument, "start height should not exceed tip height")))
+		return res, nil
```
Comment (CoderZhi) [inline api/coreservice.go:1045]: `return nil, nil`

File: api/grpcserver.go [hunk 8]
```diff
@@ -162,6 +162,9 @@ func (svr *GRPCServer) GetBlockMetas(ctx context.Context, in *iotexapi.GetBlockM
 	case in.GetByIndex() != nil:
 		request := in.GetByIndex()
 		ret, err = svr.coreService.BlockMetas(request.Start, request.Count)
+		if len(ret) == 0 {
```
Comment (CoderZhi) [inline api/grpcserver.go:165]: don't overwrite err if err != nil

File: api/web3server.go [hunk 9]
```diff
@@ -98,9 +103,10 @@ var (
 	errInvalidFiterID = errors.New("filter not found")
 	errInvalidBlock   = errors.New("invalid block")
 
-	pendingBlockNumber  = "pending"
-	latestBlockNumber   = "latest"
-	earliestBlockNumber = "earliest"
+	_emptyResult        *struct{} = nil // null is returned when no result was found
```
Comment (CoderZhi) [inline api/web3server.go:106]: _emptyResult is used in multiple places as nil interface{}, it makes code harder to read. For example, in function getBlockByNumber, what's an empty result? nil or &Block{}?

Comment (Liuhaai) [reply api/web3server.go]: good

**Review by dustinxie — APPROVED** [review]

**Review by raullenchai — APPROVED** [review]

**Pull request merged**